finat/quadrature_element.py

```diff
@@ -99,18 +104,24 @@ def point_evaluation(self, order, refcoords, entity=None):
     def dual_basis(self):
         # Point evaluations at each quadrature point with no derivatives
         duals = []
-        for pt in self._rule.point_set.points:
-            point_set = PointSingleton(pt)
+        if isinstance(self._rule.point_set, UnknownPointSingleton):
+            point_set = self._rule.point_set
```

Comment:
I think this code is hamstrung by the lack of actual data structure for the duals.

I think you can avoid this special casing if you change the iteration over the points to be something like:


```
points = gem.ComponentTensor(self._rule.point_set.expression, self._rule.indices)
for idx in numpy.ndindex(points.shape):
    pt = gem.partial_indexed(points, idx)
    duals.append(pt ...)
```
Or does the duals "data structure" need something different there?